onnxruntime/test/providers/provider_test_utils.h

```diff
@@ -422,18 +433,43 @@ class OpTester {
   enum class ExpectResult { kExpectSuccess,
                             kExpectFailure };
 
+  using CustomOutputVerifierFn =
+      std::function<void(const std::vector<OrtValue>& /*fetches*/, const std::string& /*provider_type*/)>;
+
   void Run(ExpectResult expect_result = ExpectResult::kExpectSuccess, const std::string& expected_failure_string = "",
            const std::unordered_set<std::string>& excluded_provider_types = {},
            const RunOptions* run_options = nullptr,
            std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers = nullptr,
-           ExecutionMode execution_mode = ExecutionMode::ORT_SEQUENTIAL);
+           ExecutionMode execution_mode = ExecutionMode::ORT_SEQUENTIAL,
+           const CustomOutputVerifierFn& custom_output_verifier = {});
 
   void Run(SessionOptions session_options,
            ExpectResult expect_result = ExpectResult::kExpectSuccess,
            const std::string& expected_failure_string = "",
            const std::unordered_set<std::string>& excluded_provider_types = {},
            const RunOptions* run_options = nullptr,
-           std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers = nullptr);
+           std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers = nullptr,
+           const CustomOutputVerifierFn& custom_output_verifier = {});
+
+  std::vector<MLValue> GetFetches() { return fetches_; }
+
+  std::unique_ptr<onnxruntime::Model> BuildGraph(const std::unordered_map<std::string, int>& extra_domain_to_version = {});
+
+  // storing p_model as cache
+  void SetModelCache(std::shared_ptr<onnxruntime::Model> model) {
+     cached_model_ = model;
+  }
+
+  std::shared_ptr<onnxruntime::Model> GetModelCache() {
+    return cached_model_;
+  }
+
+  // clear input/output data, fetches will be cleared in Run()
+  void ClearData() {
+    input_data_.clear();
+    output_data_.clear();
+    initializer_index_.clear();
+  }
 
```

Comment:
Generally it's cleaner to create a new instance instead of having a 'Clear' method. Is there a reason why that approach couldn't be used?

Reply:
We want to maintained the graph state to be re-used in loops. Agreed it's cleaner to create new instance. however, with the cached graph, tests time is faster (4~8% improvement). This part is only used by gradient tests, will not impact others.